File: apps/ios/ExpCore/Sources/API/CodingSessionsApi.swift
import Foundation

/// Input for `codingSessions.mergePr` (EXP-734).
public struct MergeSessionPrInput: Encodable, Sendable {
    public let sessionId: String

    public init(sessionId: String) {
        self.sessionId = sessionId
    }
}

/// EXP-734: the run's own pull request — the one an action or chat run opened
/// through MCP `exponential_pr_open({repositoryId, head})`, which links no
/// issue at all. Issue and batch runs merge through `IssuesApi.mergePr`.
public final class CodingSessionsApi: Sendable {
    private let trpc: TrpcClient

    public init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    /// Squash-merge the session's own PR via the GitHub App. Nothing is
    /// written locally: the server flips the row's `pr_state` to `merged` (and
    /// ends the run unless the team keeps sessions on merge), and both land
    /// here through Electric sync.
    public func mergePr(accountId: String, sessionId: String) async throws {
        try await trpc.mutationVoid(
            accountId: accountId,
            path: "codingSessions.mergePr",
            input: MergeSessionPrInput(sessionId: sessionId)
        )
    }
}

File: apps/android/app/src/main/java/com/exponential/app/data/api/CodingSessionsApi.kt
package com.exponential.app.data.api

import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * `codingSessions.mergePr` (EXP-734): the counterpart of `issues.mergePr` for
 * a run whose pull request links NO issue — an action or chat run that opened
 * one via MCP `exponential_pr_open({repositoryId, head})`. Nothing is
 * completed or moved: the server merges the PR, flips the session row's
 * `pr_state` to `merged` and (unless the team keeps sessions on merge) ends
 * the run, all of which arrive through Electric sync.
 */
@Serializable
data class MergeSessionPrInput(
    @SerialName("sessionId") val sessionId: String,
)

@Singleton
class CodingSessionsApi @Inject constructor(private val trpc: TrpcClient) {

    /**
     * Squash-merge the run's own PR via the GitHub App. Errors match
     * `issues.mergePr` (conflicts, branch protection, GitHub App failures), so
     * callers keep rendering them through [com.exponential.app.domain.MergeFailure].
     */
    suspend fun mergePr(accountId: String, sessionId: String) {
        trpc.mutationUnit(
            accountId,
            path = "codingSessions.mergePr",
            input = MergeSessionPrInput(sessionId),
            inputSerializer = MergeSessionPrInput.serializer(),
        )
    }
}
